src/components/sunCard/SunCard.ts

```diff
@@ -0,0 +1,244 @@
+import { HomeAssistant } from 'custom-card-helpers'
+import { CSSResult, customElement, LitElement, state, TemplateResult } from 'lit-element'
+import cardStyles from '../../cardStyles'
+import { Constants } from '../../constants'
+import { ESunCardErrors, ISunCardConfig, TSunCardData, TSunCardTimes, TSunInfo } from '../../types'
+import { HelperFunctions } from '../../utils/HelperFunctions'
+import { I18N } from '../../utils/I18N'
+import { SunCardContent } from './SunCardContent'
+import { SunCardEditor } from '../sunCardEditor'
+import { SunErrorContent } from '../SunErrorContent'
+
+@customElement('sun-card')
+export class SunCard extends LitElement {
+  static readonly cardType = 'sun-card'
+  static readonly cardName = 'Sun Card'
+  static readonly cardDescription = 'Custom card that display a graph to track the sun position and related events'
+
+  @state()
+  private config: ISunCardConfig = {
+    type: SunCard.cardType
+  }
+
+  @state()
+  private data!: TSunCardData
+
+  private hasRendered = false
+  private lastHass!: HomeAssistant
+
+  set hass (hass: HomeAssistant) {
+    this.lastHass = hass
+
+    if (!this.hasRendered) {
+      this.populateConfigFromHass()
+      return
+    }
+
+    this.processLastHass()
+  }
+
+  static getConfigElement (): HTMLElement {
+    return document.createElement(SunCardEditor.cardType)
+  }
+
+  // called by HASS whenever config changes
+  public setConfig (config: ISunCardConfig): void {
+    const newConfig = { ...this.config }
+    newConfig.title = config.title
+    newConfig.darkMode = config.darkMode
+    newConfig.language = config.language
+    newConfig.use12hourClock = config.use12hourClock
+    newConfig.component = config.component ?? Constants.DEFAULT_CONFIG.component
+    
+    if (this.config.language && !HelperFunctions.isValidLanguage(this.config.language)) {
+      throw Error(`${config.language} is not a supported language. Supported languages: ${Constants.LOCALIZATION_LANGUAGES}`)
+    }
+
+    const defaultFields = Constants.DEFAULT_CONFIG.fields!
+
+    newConfig.fields = {
+      sunrise: config.fields?.sunrise ?? defaultFields.sunrise,
+      sunset: config.fields?.sunset ?? defaultFields.sunset,
+
+      dawn: config.fields?.dawn ?? defaultFields.dawn,
+      noon: config.fields?.noon ?? defaultFields.noon,
+      dusk: config.fields?.dusk ?? defaultFields.dusk,
+
+      azimuth: config.fields?.azimuth ?? defaultFields.azimuth,
+      elevation: config.fields?.elevation ?? defaultFields.elevation
+    }
+
+    this.config = newConfig
+    if (this.lastHass) {
+      this.populateConfigFromHass()
+    }
+  }
```

Comment:
These changes were required to allow the deletion of some configurations with the visual editor but also to fix how lit element works with objects (making it reactive to property changes adds more complexity than just reassign to a new object)